src/server/worker/datum/iterator.go

```diff
@@ -39,6 +40,73 @@ func NewIterator(ctx context.Context, c pfs.APIClient, taskDoer task.Doer, input
 	return NewFileSetIterator(ctx, c, fileSetID, nil), nil
 }
 
+func NewCreateDatumStreamIterator(ctx context.Context, c pfs.APIClient, taskDoer task.Doer, input *pps.Input) Iterator {
+	ctx = pctx.Child(ctx, "CreateDatumStream")
+	cds := &createDatumStream{
+		ctx:               ctx,
+		c:                 c,
+		taskDoer:          taskDoer,
+		input:             input,
+		requestDatumsChan: make(chan bool),
+		fsidChan:          make(chan string),
+		errChan:           make(chan error, 1),
+	}
+	go cds.create(cds.input, cds.fsidChan)
+	return &createDatumStreamIterator{
+		createDatumStream: cds,
+		metaBuffer:        make([]*Meta, 0),
+	}
+}
+
+type createDatumStreamIterator struct {
+	*createDatumStream
+	metaBuffer []*Meta
+}
+
+// Return value of nil means all datums have been returned.
+// Return value of errutil.ErrBreak means that iteration is paused and
+// will be resumed when client asks for more.
+// Return value of anything else means an error.
+func (it *createDatumStreamIterator) Iterate(cb func(*Meta) error) error {
+	for {
+		// Consume leftover datums from the buffer first before asking
+		// to create more.
+		if len(it.metaBuffer) == 0 {
+			select {
+			case it.requestDatumsChan <- true:
+			case fsid, ok := <-it.fsidChan:
+				if !ok {
+					return nil
+				}
+				fsidIt := NewFileSetIterator(it.ctx, it.c, fsid, nil)
+				if err := fsidIt.Iterate(func(meta *Meta) error {
+					err := cb(meta)
+					// If callback has consumed all the datums it needs,
+					// add the leftover datums to the buffer.
+					if errors.Is(err, errutil.ErrBreak) {
```

Comment:
Check for nil before calling errors.Is;  example of how it can go wrong: https://go.dev/play/p/dVK0okSjzR5

Reply:
Is more than a simple nil check needed? https://go.dev/play/p/6bJViY9Rg4y still errors.

Reply:
No, I think all you need is a check.  If err is nil, don't call Is... something else is happening in that example (the return value is nil but a non-nil error, so doesn't actually demonstrate this problem)